algebra-utils/src/msm/variable_base.rs

```diff
@@ -0,0 +1,603 @@
+        let fr_one = G::ScalarField::one().into_repr();
+
+        let zero = G::zero().into_projective();
+        let window_starts: Vec<_> = (0..num_bits).step_by(c).collect();
+
+        // Each window is of size `c`.
+        // We divide up the bits 0..num_bits into windows of size `c`, and
+        // in parallel process each such window.
+        let window_sums: Vec<_> = window_starts
+            .into_par_iter()
+            .map(|w_start| {
+                let mut res = zero;
+                // We don't need the "zero" bucket, so we only have 2^c - 1 buckets
+                let mut buckets = vec![zero; (1 << c) - 1];
+                scalars.iter().zip(bases).filter(|(s, _)| !s.is_zero()).for_each(|(&scalar, base)|  {
+                    if scalar == fr_one {
+                        // We only process unit scalars once in the first window.
+                        if w_start == 0 {
+                            res.add_assign_mixed(base);
+                        }
+                    } else {
+                        let mut scalar = scalar;
+
+                        // We right-shift by w_start, thus getting rid of the
+                        // lower bits.
+                        scalar.divn(w_start as u32);
+
+                        // We mod the remaining bits by the window size.
+                        let scalar = scalar.as_ref()[0] % (1 << c);
+
+                        // If the scalar is non-zero, we update the corresponding
+                        // bucket.
+                        // (Recall that `buckets` doesn't have a zero bucket.)
+                        if scalar != 0 {
+                            buckets[(scalar - 1) as usize].add_assign_mixed(&base);
+                        }
+                    }
+                });
+                G::Projective::batch_normalization(&mut buckets);
+
+                let mut running_sum = G::Projective::zero();
+                for b in buckets.into_iter().map(|g| g.into_affine()).rev() {
+                    running_sum.add_assign_mixed(&b);
+                    res += &running_sum;
+                }
+
+                res
+            })
+            .collect();
+
+        // We store the sum for the lowest window.
+        let lowest = window_sums.first().unwrap();
+
+        // We're traversing windows from high to low.
+        window_sums[1..].iter().rev().fold(zero, |mut total, sum_i| {
+            total += sum_i;
+            for _ in 0..c {
+                total.double_in_place();
+            }
+            total
+        }) + lowest
+    }
+
+    fn msm_inner_cpu<G>(
+        bases: &[G],
+        scalars: &[<G::ScalarField as PrimeField>::BigInt]
+    ) -> G::Projective
+    where
+        G: AffineCurve,
+        G::Projective: ProjectiveCurve<Affine = G>
+    {
+        let scal_len = scalars.len();
+
+        if TypeId::of::<G>() == TypeId::of::<algebra::curves::bn_382::G1Affine>()
```

Comment:
Let's remove the hardcoding of all these window sizes: it causes compilation issues as it is now (it's fixable but these parameters are not really very useful, so let's just keep with the default formula to compute the window size)